keps/sig-storage/20200120-generic-data-populators.md

```diff
@@ -0,0 +1,307 @@
+the image repository, and to allow creation of PVCs from these custom resources such
+that the volumes come pre-populated with the correct data. Efficient population of the
+data could be left up to a purpose-built controller that knows how to get the bits
+where they need to be with minimal I/O.
+
+#### Backup/Restore
+
+Without getting into the details of how backup/restore should be implemented, it's
+clear that whatever design one chooses, a necessary step is to have the user
+(or higher level controller) create a PVC that points to the backup they want to
+restore, and have the data appear in that volume somehow.
+
+One can imagine backups simply being a special case of snapshots, in which case the
+existing design is sufficient, but if you want anything more sophisticated, there
+will inevitably be a new object type that represents a backup. While it's arguable
+that backup should be something CSI plugins should be exclusively responsible for,
+one can also argue that generic backup tools should also exist which can backup
+and restore all kind of volumes. Those tools will be apart from CSI plugins and
+yet need a way to populate volumes with restored volumes.
+
+It's also likely that multiple backup/restore implementations will be developed,
+and it's not a good idea to pick a winner at the Kubernetes API layer. It makes
+more sense to enable developers to try different approaches by making the API allow
+restoring from various kinds of things. 
+
+### Implementation Details/Notes/Constraints
+
+As noted above, the proposal is extremely simple -- just remove the validation on
+the `DataSource` field. This raises the question of WHAT will happen when users
+put new things in that field, and HOW populators will actually work with so small
+a change.
+
+It's first important to note that only consumers of the `DataSource` field today
+are the various dynamic provisioners, most notably the external-provisioner CSI
+sidecar. If the external-provisioner sidecar sees a data source it doesn't
+understand, it simply ignores the request, which is both important for forward
+compatibility, and also perfect for the purposes of a data populator. This allows
+developers to add new types of data sources that the dynamic provisioners will
+simply ignore, enabling a different controller to see these objects and respond
+to them.
+
+The way I expect most data populators to work in practice is that they will see
+a request for a PVC `foo` with a data source of `bar`. The populator will create
+a shadow PVC `foo'` in another namespace with all of the same spec except the
+data source will be empty. The dynamic provisioner will ignore `foo` but will
+respond to `foo'` by creating an empty PV. The populator can create a pod
+attached to `foo'` and fill it with data pointed to by `bar` using whatever
+mechanism the developer chooses. After the volume is populated, the pod can be
+deleted, and the PV can be rebound to `foo`.
```

Comment:
We don't have a general model of a PV being re-bound, do we?

Reply:
The only case I know of this being done is for non-dynamic PVs when they get recycled. It's not a common thing, but I think it should be for the populator case.